pallets/liquidity-pools-gateway/src/message_processing.rs

```diff
@@ -0,0 +1,489 @@
+use cfg_primitives::LP_DEFENSIVE_WEIGHT;
+use cfg_traits::liquidity_pools::{
+	InboundMessageHandler, LPEncoding, MessageQueue, MessageSender, Proof, RouterProvider,
+};
+use cfg_types::domain_address::{Domain, DomainAddress};
+use frame_support::{
+	dispatch::DispatchResult,
+	ensure,
+	pallet_prelude::{Decode, Encode, Get, TypeInfo},
+	weights::Weight,
+};
+use parity_scale_codec::MaxEncodedLen;
+use sp_arithmetic::traits::{EnsureAddAssign, EnsureSub};
+use sp_runtime::DispatchError;
+
+use crate::{
+	message::GatewayMessage, Config, Error, Pallet, PendingInboundEntries, Routers, SessionIdStore,
+};
+
+/// Type used when storing inbound message information.
+#[derive(Debug, Encode, Decode, Clone, Eq, MaxEncodedLen, PartialEq, TypeInfo)]
+#[scale_info(skip_type_params(T))]
+pub enum InboundEntry<T: Config> {
+	Message {
+		session_id: T::SessionId,
+		domain_address: DomainAddress,
+		message: T::Message,
+		expected_proof_count: u32,
+	},
+	Proof {
+		session_id: T::SessionId,
+		current_count: u32,
+	},
+}
+
+/// Type used when processing inbound messages.
+#[derive(Clone)]
+pub struct InboundProcessingInfo<T: Config> {
+	pub domain_address: DomainAddress,
+	pub router_ids: Vec<T::RouterId>,
+	pub current_session_id: T::SessionId,
+	pub expected_proof_count_per_message: u32,
+}
+
+impl<T: Config> Pallet<T> {
+	/// Retrieves all available routers for a domain and then filters them based
+	/// on the routers that we have in storage.
+	pub fn get_router_ids_for_domain(domain: Domain) -> Result<Vec<T::RouterId>, DispatchError> {
+		let all_routers_for_domain = T::RouterProvider::routers_for_domain(domain);
+
+		let stored_routers = Routers::<T>::get();
+
+		let res = all_routers_for_domain
+			.iter()
+			.filter(|x| stored_routers.iter().any(|y| *x == y))
+			.map(|x| x.clone())
+			.collect::<Vec<_>>();
+
+		if res.is_empty() {
+			return Err(Error::<T>::NotEnoughRoutersForDomain.into());
+		}
+
+		Ok(res)
+	}
+
+	/// Calculates and returns the proof count required for processing one
+	/// inbound message.
+	fn get_expected_proof_count(router_ids: &Vec<T::RouterId>) -> Result<u32, DispatchError> {
+		let expected_proof_count = router_ids
+			.len()
+			.ensure_sub(1)
+			.map_err(|_| Error::<T>::NotEnoughRoutersForDomain)?;
+
+		Ok(expected_proof_count as u32)
```

Comment:
Nit: Use safe math. This will be flagged in audit, I presume.
```suggestion
		Ok(expected_proof_count.saturated_into())
```

Reply:
Damn, I always forget about these, will fix. Thank you!